File: python/db/connector.py
from __future__ import annotations

import os

import boto3
from boto3.dynamodb.conditions import Key

from db.base import BaseConnector, ConflictError, Ok, Result

TABLE_NAME = os.environ.get("RAILBACK_DDB_TABLE", "RailBack")
REGION = "eu-north-1"
ENDPOINT_URL = os.environ.get("DYNAMODB_ENDPOINT_URL", None)

_ADMIN_STRIPPED_FIELDS = frozenset({"iban_enc", "bic_enc"})


def _is_plain_ticket_sk(sk: str) -> bool:
    return sk.startswith("TICKET#") and "#" not in sk[len("TICKET#"):]


def _table_resource(table=None):
    if table is not None:
        return table
    kwargs = {}
    if ENDPOINT_URL:
        kwargs["endpoint_url"] = ENDPOINT_URL
        kwargs["aws_access_key_id"] = "fake"
        kwargs["aws_secret_access_key"] = "fake"
    ddb = boto3.resource("dynamodb", region_name=REGION, **kwargs)
    return ddb.Table(TABLE_NAME)


# ---------------------------------------------------------------------------
# User  —  pk=USER#{email}  sk=PROFILE
# ---------------------------------------------------------------------------

class UserConnector(BaseConnector):
    def get(self, email: str) -> Result:
        return self._get(f"USER#{email}", "PROFILE")

    def get_for_admin(self, email: str) -> Result:
        result = self._get(f"USER#{email}", "PROFILE")
        if result.is_err():
            return result
        item = result.unwrap()
        if item:
            item = {k: v for k, v in item.items() if k not in _ADMIN_STRIPPED_FIELDS}
        return Ok(item)

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", "PROFILE", updates)

    def list_all(self, limit: int | None = None) -> Result:
        return self._query(IndexName="gsi1", KeyConditionExpression=Key("gsi1_pk").eq("USER"), **({"Limit": limit} if limit is not None else {}))


# ---------------------------------------------------------------------------
# Admin  —  pk=ADMIN#{email}  sk=PROFILE
# ---------------------------------------------------------------------------

class AdminConnector(BaseConnector):
    def get(self, email: str) -> Result:
        return self._get(f"ADMIN#{email}", "PROFILE")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, updates: dict) -> Result:
        return self._update_fields(f"ADMIN#{email}", "PROFILE", updates)

    def list_all(self, limit: int | None = None) -> Result:
        return self._query(IndexName="gsi1", KeyConditionExpression=Key("gsi1_pk").eq("ADMIN"), **({"Limit": limit} if limit is not None else {}))


# ---------------------------------------------------------------------------
# Ticket  —  pk=USER#{email}  sk=TICKET#{ticket_id}
# ---------------------------------------------------------------------------

class TicketConnector(BaseConnector):
    def get(self, email: str, ticket_id: str) -> Result:
        return self._get(f"USER#{email}", f"TICKET#{ticket_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, ticket_id: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", f"TICKET#{ticket_id}", updates)

    def list_for_user(self, email: str, limit: int | None = None) -> Result:
        result = self._query(
            KeyConditionExpression=Key("pk").eq(f"USER#{email}") & Key("sk").begins_with("TICKET#"),
        )
        if result.is_err():
            return result
        items = [i for i in result.unwrap() if _is_plain_ticket_sk(i["sk"])]
        return Ok(items[:limit] if limit is not None else items)

    def get_by_train(self, train_nr: str, date: str, limit: int | None = None) -> Result:
        return self._query(
            IndexName="gsi1",
            KeyConditionExpression=Key("gsi1_pk").eq(f"TRAIN#{train_nr}#{date}"),
            **({"Limit": limit} if limit is not None else {}),
        )

    def check_barcode_duplicate(self, barcode_uid: str) -> Result:
        result = self._query(
            IndexName="gsi2",
            KeyConditionExpression=Key("gsi2_pk").eq("BARCODE") & Key("gsi2_sk").eq(barcode_uid),
            Limit=1,
        )
        if result.is_err():
            return result
        items = result.unwrap()
        return Ok(items[0] if items else None)

    def list_email_pending(self, limit: int) -> Result:
        return self._query(
            IndexName="gsi_email_pending",
            KeyConditionExpression=Key("gsi_email_pending_pk").eq("EMAIL_PENDING"),
            ScanIndexForward=True,
            Limit=limit,
        )


# ---------------------------------------------------------------------------
# TicketOwner  —  pk=TICKET#{ticket_id}  sk=OWNER
# ---------------------------------------------------------------------------

class TicketOwnerConnector(BaseConnector):
    def get(self, ticket_id: str) -> Result:
        return self._get(f"TICKET#{ticket_id}", "OWNER")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, ticket_id: str, updates: dict) -> Result:
        return self._update_fields(f"TICKET#{ticket_id}", "OWNER", updates)


# ---------------------------------------------------------------------------
# RawUpload  —  pk=USER#{email}  sk=RAW#{ticket_id}
# ---------------------------------------------------------------------------

class RawUploadConnector(BaseConnector):
    def get(self, email: str, ticket_id: str) -> Result:
        return self._get(f"USER#{email}", f"RAW#{ticket_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, ticket_id: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", f"RAW#{ticket_id}", updates)


# ---------------------------------------------------------------------------
# RenderedPdf  —  pk=USER#{email}  sk=RENDERED#{ticket_id}
# ---------------------------------------------------------------------------

class RenderedPdfConnector(BaseConnector):
    def get(self, email: str, ticket_id: str) -> Result:
        return self._get(f"USER#{email}", f"RENDERED#{ticket_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, ticket_id: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", f"RENDERED#{ticket_id}", updates)


# ---------------------------------------------------------------------------
# OriginalReceipt  —  pk=USER#{email}  sk=TICKET#{ticket_id}#BELEG#{beleg_id}
# ---------------------------------------------------------------------------

class OriginalReceiptConnector(BaseConnector):
    def get(self, email: str, ticket_id: str, beleg_id: str) -> Result:
        return self._get(f"USER#{email}", f"TICKET#{ticket_id}#BELEG#{beleg_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, ticket_id: str, beleg_id: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", f"TICKET#{ticket_id}#BELEG#{beleg_id}", updates)

    def list_for_ticket(self, email: str, ticket_id: str, limit: int | None = None) -> Result:
        return self._query(
            KeyConditionExpression=Key("pk").eq(f"USER#{email}") & Key("sk").begins_with(f"TICKET#{ticket_id}#BELEG#"),
            **({"Limit": limit} if limit is not None else {}),
        )


# ---------------------------------------------------------------------------
# SepaMandate  —  pk=USER#{email}  sk=TICKET#{ticket_id}#MANDATE
# ---------------------------------------------------------------------------

class SepaMandateConnector(BaseConnector):
    def get(self, email: str, ticket_id: str) -> Result:
        return self._get(f"USER#{email}", f"TICKET#{ticket_id}#MANDATE")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, ticket_id: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", f"TICKET#{ticket_id}#MANDATE", updates)

    def stamp_pain008_built(self, email: str, ticket_id: str, batch_id: str, s3_key: str, built_at: str) -> Result:
        return self._update_if(
            f"USER#{email}", f"TICKET#{ticket_id}#MANDATE",
            {"pain008_built_at": built_at, "pain008_batch_id": batch_id, "pain008_s3_key": s3_key},
            "attribute_exists(pk) AND attribute_not_exists(pain008_built_at)",
        )


# ---------------------------------------------------------------------------
# SepaReport  —  pk=SEPA#REPORT#{date}  sk=REPORT#{report_id}
# ---------------------------------------------------------------------------

class SepaReportConnector(BaseConnector):
    def get(self, date: str, report_id: str) -> Result:
        return self._get(f"SEPA#REPORT#{date}", f"REPORT#{report_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, date: str, report_id: str, updates: dict) -> Result:
        return self._update_fields(f"SEPA#REPORT#{date}", f"REPORT#{report_id}", updates)

    def list_by_date(self, date: str, limit: int | None = None) -> Result:
        return self._query(KeyConditionExpression=Key("pk").eq(f"SEPA#REPORT#{date}"), **({"Limit": limit} if limit is not None else {}))


# ---------------------------------------------------------------------------
# TrainSegmentDelay  —  pk=TRAIN#{train_nr}#{date}  sk=SEG#{seg_id}
# ---------------------------------------------------------------------------

class TrainSegmentDelayConnector(BaseConnector):
    def get(self, train_nr: str, date: str, seg_id: str) -> Result:
        return self._get(f"TRAIN#{train_nr}#{date}", f"SEG#{seg_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, train_nr: str, date: str, seg_id: str, updates: dict) -> Result:
        return self._update_fields(f"TRAIN#{train_nr}#{date}", f"SEG#{seg_id}", updates)

    def list_for_train(self, train_nr: str, date: str, limit: int | None = None) -> Result:
        return self._query(
            KeyConditionExpression=Key("pk").eq(f"TRAIN#{train_nr}#{date}") & Key("sk").begins_with("SEG#"),
            **({"Limit": limit} if limit is not None else {}),
        )

    def route_lookup(self, origin_eva: int, date: str, from_time: str, to_time: str, limit: int | None = None) -> Result:
        return self._query(
            IndexName="gsi1",
            KeyConditionExpression=Key("gsi1_pk").eq(f"STATION#{origin_eva}#{date}") & Key("gsi1_sk").between(from_time, to_time + "￿"),
            **({"Limit": limit} if limit is not None else {}),
        )


# ---------------------------------------------------------------------------
# RouteTemplate  —  pk=USER#{email}  sk=TEMPLATE#{template_id}
# ---------------------------------------------------------------------------

class RouteTemplateConnector(BaseConnector):
    def get(self, email: str, template_id: str) -> Result:
        return self._get(f"USER#{email}", f"TEMPLATE#{template_id}")

    def put(self, item: dict) -> Result:
        return self._put(item)

    def update(self, email: str, template_id: str, updates: dict) -> Result:
        return self._update_fields(f"USER#{email}", f"TEMPLATE#{template_id}", updates)

    def list_for_user(self, email: str, limit: int | None = None) -> Result:
        return self._query(
            KeyConditionExpression=Key("pk").eq(f"USER#{email}") & Key("sk").begins_with("TEMPLATE#"),
            **({"Limit": limit} if limit is not None else {}),
        )


# ---------------------------------------------------------------------------
# Root connector
# ---------------------------------------------------------------------------

class RailBackConnector:
    def __init__(self, table=None):
        self._table = _table_resource(table)
        self.user = UserConnector(self._table)
        self.admin = AdminConnector(self._table)
        self.ticket = TicketConnector(self._table)
        self.ticket_owner = TicketOwnerConnector(self._table)
        self.raw_upload = RawUploadConnector(self._table)
        self.rendered_pdf = RenderedPdfConnector(self._table)
        self.receipt = OriginalReceiptConnector(self._table)
        self.mandate = SepaMandateConnector(self._table)
        self.sepa_report = SepaReportConnector(self._table)
        self.train_delay = TrainSegmentDelayConnector(self._table)
        self.route_template = RouteTemplateConnector(self._table)

    def delete_user(self, email: str) -> Result:
        all_items = self.user._query(KeyConditionExpression=Key("pk").eq(f"USER#{email}"))
        if all_items.is_err():
            return all_items
        items = all_items.unwrap()
        keys = [(item["pk"], item["sk"]) for item in items]
        for item in items:
            sk = item["sk"]
            if _is_plain_ticket_sk(sk):
                keys.append((f"TICKET#{sk[len('TICKET#'):]}", "OWNER"))
        return self.user._batch_delete(keys)

    def delete_ticket(self, email: str, ticket_id: str) -> Result:
        children = self.ticket._query(
            KeyConditionExpression=Key("pk").eq(f"USER#{email}") & Key("sk").begins_with(f"TICKET#{ticket_id}#")
        )
        if children.is_err():
            return children
        keys = [(item["pk"], item["sk"]) for item in children.unwrap()]
        keys += [
            (f"USER#{email}", f"TICKET#{ticket_id}"),
            (f"TICKET#{ticket_id}", "OWNER"),
            (f"USER#{email}", f"RAW#{ticket_id}"),
            (f"USER#{email}", f"RENDERED#{ticket_id}"),
        ]
        return self.ticket._batch_delete(keys)

    def delete_admin(self, email: str) -> Result:
        return self.admin._delete(f"ADMIN#{email}", "PROFILE")

    def delete_ticket_owner(self, ticket_id: str) -> Result:
        return self.ticket_owner._delete(f"TICKET#{ticket_id}", "OWNER")

    def delete_raw_upload(self, email: str, ticket_id: str) -> Result:
        return self.raw_upload._delete(f"USER#{email}", f"RAW#{ticket_id}")

    def delete_rendered_pdf(self, email: str, ticket_id: str) -> Result:
        return self.rendered_pdf._delete(f"USER#{email}", f"RENDERED#{ticket_id}")

    def delete_receipt(self, email: str, ticket_id: str, beleg_id: str) -> Result:
        return self.receipt._delete(f"USER#{email}", f"TICKET#{ticket_id}#BELEG#{beleg_id}")

    def delete_mandate(self, email: str, ticket_id: str) -> Result:
        return self.mandate._delete(f"USER#{email}", f"TICKET#{ticket_id}#MANDATE")

    def delete_sepa_report(self, date: str, report_id: str) -> Result:
        return self.sepa_report._delete(f"SEPA#REPORT#{date}", f"REPORT#{report_id}")

    def delete_train_delay(self, train_nr: str, date: str, seg_id: str) -> Result:
        return self.train_delay._delete(f"TRAIN#{train_nr}#{date}", f"SEG#{seg_id}")

    def delete_route_template(self, email: str, template_id: str) -> Result:
        return self.route_template._delete(f"USER#{email}", f"TEMPLATE#{template_id}")

File: node/src/connector.ts
import { DynamoDBDocumentClient } from "@aws-sdk/lib-dynamodb";
import { BaseConnector, ConflictError, Ok, Result, createClient, normaliseEmail } from "./base.js";
import { S3BlobConnector } from "./connectors/s3.js";

function isPlainTicketSk(sk: string): boolean {
  const tail = sk.slice("TICKET#".length);
  return sk.startsWith("TICKET#") && tail !== "" && !tail.includes("#");
}

// Email → USER# key. Trim + lowercase so mixed-case input maps into the
// same partition. See F7 (2026-07-08) — the backend layer already
// normalises via @railback/lib/storage/ddb/keys#normaliseEmail; without
// this the two systems would split partitions on any mixed-case input.
function userPk(email: string): string {
  return `USER#${normaliseEmail(email)}`;
}

function adminPk(email: string): string {
  return `ADMIN#${normaliseEmail(email)}`;
}

// ---------------------------------------------------------------------------
// User  —  pk=USER#{email}  sk=PROFILE
// ---------------------------------------------------------------------------

export class UserConnector extends BaseConnector {
  get(email: string) { return this._get(userPk(email), "PROFILE"); }
  // 2026-07-07 reversal (DECISIONS.md): admin sees the full row including
  // iban_enc / bic_enc. Decryption to plaintext happens in admin-handler
  // (backend layer) — the adapter just returns the raw row. Encryption at
  // rest is still enforced; admin authority to decrypt is a separate concern.
  getForAdmin(email: string): Promise<Result<Record<string, unknown> | null>> {
    return this._get(userPk(email), "PROFILE").then((r) => {
      if (r.isErr() || r.value === null) return r;
      return new Ok(r.value);
    });
  }
  getForAuth(email: string): Promise<Result<Record<string, unknown> | null>> {
    const norm = normaliseEmail(email);
    return this._get(userPk(norm), "PROFILE").then((r) => {
      if (r.isErr() || r.value === null) return r;
      const { hashed_password, user_state } = r.value;
      return new Ok({ email: norm, hashed_password, user_state });
    });
  }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), "PROFILE", updates);
  }
  listAll(limit?: number) {
    return this._query({ IndexName: "gsi1", KeyConditionExpression: "gsi1_pk = :v", ExpressionAttributeValues: { ":v": "USER" }, ...(limit !== undefined ? { Limit: limit } : {}) });
  }
}

// ---------------------------------------------------------------------------
// Admin  —  pk=ADMIN#{email}  sk=PROFILE
// ---------------------------------------------------------------------------

export class AdminConnector extends BaseConnector {
  get(email: string) { return this._get(adminPk(email), "PROFILE"); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, updates: Record<string, unknown>) {
    return this._updateFields(adminPk(email), "PROFILE", updates);
  }
  listAll(limit?: number) {
    return this._query({ IndexName: "gsi1", KeyConditionExpression: "gsi1_pk = :v", ExpressionAttributeValues: { ":v": "ADMIN" }, ...(limit !== undefined ? { Limit: limit } : {}) });
  }
}

// ---------------------------------------------------------------------------
// Ticket  —  pk=USER#{email}  sk=TICKET#{id}
// ---------------------------------------------------------------------------

export class TicketConnector extends BaseConnector {
  get(email: string, ticketId: string) { return this._get(userPk(email), `TICKET#${ticketId}`); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, ticketId: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), `TICKET#${ticketId}`, updates);
  }
  async listForUser(email: string, limit?: number): Promise<Result<Record<string, unknown>[]>> {
    const r = await this._query({
      KeyConditionExpression: "pk = :pk AND begins_with(sk, :prefix)",
      ExpressionAttributeValues: { ":pk": userPk(email), ":prefix": "TICKET#" },
    });
    if (r.isErr()) return r;
    const items = r.value.filter((i) => isPlainTicketSk(i["sk"] as string));
    return new Ok(limit !== undefined ? items.slice(0, limit) : items);
  }
  getByTrain(trainNr: string, date: string, limit?: number) {
    return this._query({ IndexName: "gsi1", KeyConditionExpression: "gsi1_pk = :v", ExpressionAttributeValues: { ":v": `TRAIN#${trainNr}#${date}` }, ...(limit !== undefined ? { Limit: limit } : {}) });
  }
  async checkBarcodeDuplicate(barcodeUid: string): Promise<Result<Record<string, unknown> | null>> {
    const r = await this._query({
      IndexName: "gsi2",
      KeyConditionExpression: "gsi2_pk = :pk AND gsi2_sk = :sk",
      ExpressionAttributeValues: { ":pk": "BARCODE", ":sk": barcodeUid },
      Limit: 1,
    });
    if (r.isErr()) return r;
    const hit = r.value[0];
    if (!hit) return new Ok(null);
    // GSI2 is KEYS_ONLY (DB_SCHEMA.md) — the hit carries only pk/sk/gsi2_*.
    // Refetch the full ticket so callers get email/ticket_state/etc.
    return this._get(hit["pk"] as string, hit["sk"] as string);
  }
  async listEmailPending(limit: number): Promise<Result<Record<string, unknown>[]>> {
    const r = await this._query({
      IndexName: "gsi_email_pending",
      KeyConditionExpression: "gsi_email_pending_pk = :v",
      ExpressionAttributeValues: { ":v": "EMAIL_PENDING" },
      ScanIndexForward: true,
      Limit: limit,
    });
    if (r.isErr()) return r;
    // GSI_EMAIL_PENDING is KEYS_ONLY — refetch each hit's full item by pk/sk.
    // Oldest-first order from the index is preserved.
    const full: Record<string, unknown>[] = [];
    for (const hit of r.value) {
      const one = await this._get(hit["pk"] as string, hit["sk"] as string);
      if (one.isErr()) return one as unknown as Result<Record<string, unknown>[]>;
      if (one.value !== null) full.push(one.value);
    }
    return new Ok(full);
  }
}

// ---------------------------------------------------------------------------
// TicketOwner  —  pk=TICKET#{id}  sk=OWNER
// ---------------------------------------------------------------------------

export class TicketOwnerConnector extends BaseConnector {
  get(ticketId: string) { return this._get(`TICKET#${ticketId}`, "OWNER"); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(ticketId: string, updates: Record<string, unknown>) {
    return this._updateFields(`TICKET#${ticketId}`, "OWNER", updates);
  }
}

// ---------------------------------------------------------------------------
// RawUpload  —  pk=USER#{email}  sk=RAW#{id}
// ---------------------------------------------------------------------------

export class RawUploadConnector extends BaseConnector {
  get(email: string, ticketId: string) { return this._get(userPk(email), `RAW#${ticketId}`); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, ticketId: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), `RAW#${ticketId}`, updates);
  }
}

// ---------------------------------------------------------------------------
// RenderedPdf  —  pk=USER#{email}  sk=RENDERED#{id}
// ---------------------------------------------------------------------------

export class RenderedPdfConnector extends BaseConnector {
  get(email: string, ticketId: string) { return this._get(userPk(email), `RENDERED#${ticketId}`); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, ticketId: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), `RENDERED#${ticketId}`, updates);
  }
}

// ---------------------------------------------------------------------------
// OriginalReceipt  —  pk=USER#{email}  sk=TICKET#{id}#BELEG#{id}
// ---------------------------------------------------------------------------

export class OriginalReceiptConnector extends BaseConnector {
  get(email: string, ticketId: string, belegId: string) {
    return this._get(userPk(email), `TICKET#${ticketId}#BELEG#${belegId}`);
  }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, ticketId: string, belegId: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), `TICKET#${ticketId}#BELEG#${belegId}`, updates);
  }
  listForTicket(email: string, ticketId: string, limit?: number) {
    return this._query({
      KeyConditionExpression: "pk = :pk AND begins_with(sk, :prefix)",
      ExpressionAttributeValues: { ":pk": userPk(email), ":prefix": `TICKET#${ticketId}#BELEG#` },
      ...(limit !== undefined ? { Limit: limit } : {}),
    });
  }
}

// ---------------------------------------------------------------------------
// SepaMandate  —  pk=USER#{email}  sk=TICKET#{id}#MANDATE
// ---------------------------------------------------------------------------

export class SepaMandateConnector extends BaseConnector {
  get(email: string, ticketId: string) {
    return this._get(userPk(email), `TICKET#${ticketId}#MANDATE`);
  }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, ticketId: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), `TICKET#${ticketId}#MANDATE`, updates);
  }
  stampPain008Built(email: string, ticketId: string, batchId: string, s3Key: string, builtAt: string) {
    return this._updateIf(
      userPk(email), `TICKET#${ticketId}#MANDATE`,
      { pain008_built_at: builtAt, pain008_batch_id: batchId, pain008_s3_key: s3Key },
      "attribute_exists(pk) AND attribute_not_exists(pain008_built_at)",
    );
  }
}

// ---------------------------------------------------------------------------
// SepaReport  —  pk=SEPA#REPORT#{date}  sk=REPORT#{id}
// ---------------------------------------------------------------------------

export class SepaReportConnector extends BaseConnector {
  get(date: string, reportId: string) { return this._get(`SEPA#REPORT#${date}`, `REPORT#${reportId}`); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(date: string, reportId: string, updates: Record<string, unknown>) {
    return this._updateFields(`SEPA#REPORT#${date}`, `REPORT#${reportId}`, updates);
  }
  listByDate(date: string, limit?: number) {
    return this._query({ KeyConditionExpression: "pk = :v", ExpressionAttributeValues: { ":v": `SEPA#REPORT#${date}` }, ...(limit !== undefined ? { Limit: limit } : {}) });
  }
}

// ---------------------------------------------------------------------------
// TrainSegmentDelay  —  pk=TRAIN#{nr}#{date}  sk=SEG#{id}
// ---------------------------------------------------------------------------

export class TrainSegmentDelayConnector extends BaseConnector {
  get(trainNr: string, date: string, segId: string) { return this._get(`TRAIN#${trainNr}#${date}`, `SEG#${segId}`); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(trainNr: string, date: string, segId: string, updates: Record<string, unknown>) {
    return this._updateFields(`TRAIN#${trainNr}#${date}`, `SEG#${segId}`, updates);
  }
  listForTrain(trainNr: string, date: string, limit?: number) {
    return this._query({
      KeyConditionExpression: "pk = :pk AND begins_with(sk, :prefix)",
      ExpressionAttributeValues: { ":pk": `TRAIN#${trainNr}#${date}`, ":prefix": "SEG#" },
      ...(limit !== undefined ? { Limit: limit } : {}),
    });
  }
  routeLookup(originEva: number, date: string, fromTime: string, toTime: string, limit?: number) {
    // Route-lookup rides GSI3 (STATION#<eva>#<date> / <plannedDeparture>#<trainNr>),
    // projection ALL — DB_SCHEMA.md "GSI3". NOT gsi1 (that's the train /
    // admin-enumeration index and never carries STATION# keys).
    //
    // The ingest-delays poller writes gsi3_sk as `<date>T<HH:MM>#<trainNr>`
    // (full ISO planned_departure), not the bare `<HH:MM>#…` the older spec
    // assumed. Callers pass the window as bare HH:MM (API contract), so we
    // date-prefix the bounds to the same `<date>T<HH:MM>` shape — otherwise
    // every HH:MM bound sorts lexically below the ISO keys and BETWEEN
    // matches nothing. `date` is already pinned by gsi3_pk, so prefixing with
    // it keeps the range exact. `￿` on the upper bound sweeps the
    // `#<trainNr>` suffix.
    const from = `${date}T${fromTime}`;
    const to = `${date}T${toTime}`;
    return this._query({
      IndexName: "gsi3",
      KeyConditionExpression: "gsi3_pk = :pk AND gsi3_sk BETWEEN :from AND :to",
      ExpressionAttributeValues: { ":pk": `STATION#${originEva}#${date}`, ":from": from, ":to": to + "￿" },
      ...(limit !== undefined ? { Limit: limit } : {}),
    });
  }
}

// ---------------------------------------------------------------------------
// RouteTemplate  —  pk=USER#{email}  sk=TEMPLATE#{id}
// ---------------------------------------------------------------------------

export class RouteTemplateConnector extends BaseConnector {
  get(email: string, templateId: string) { return this._get(userPk(email), `TEMPLATE#${templateId}`); }
  put(item: Record<string, unknown>) { return this._put(item); }
  update(email: string, templateId: string, updates: Record<string, unknown>) {
    return this._updateFields(userPk(email), `TEMPLATE#${templateId}`, updates);
  }
  listForUser(email: string, limit?: number) {
    return this._query({
      KeyConditionExpression: "pk = :pk AND begins_with(sk, :prefix)",
      ExpressionAttributeValues: { ":pk": userPk(email), ":prefix": "TEMPLATE#" },
      ...(limit !== undefined ? { Limit: limit } : {}),
    });
  }
}

// ---------------------------------------------------------------------------
// Root connector
// ---------------------------------------------------------------------------

export class RailBackConnector {
  user: UserConnector;
  admin: AdminConnector;
  ticket: TicketConnector;
  ticketOwner: TicketOwnerConnector;
  rawUpload: RawUploadConnector;
  renderedPdf: RenderedPdfConnector;
  receipt: OriginalReceiptConnector;
  mandate: SepaMandateConnector;
  sepaReport: SepaReportConnector;
  trainDelay: TrainSegmentDelayConnector;
  routeTemplate: RouteTemplateConnector;
  s3blob: S3BlobConnector;

  constructor(client?: DynamoDBDocumentClient, opts?: { s3?: S3BlobConnector }) {
    const c = client ?? createClient();
    this.user = new UserConnector(c);
    this.admin = new AdminConnector(c);
    this.ticket = new TicketConnector(c);
    this.ticketOwner = new TicketOwnerConnector(c);
    this.rawUpload = new RawUploadConnector(c);
    this.renderedPdf = new RenderedPdfConnector(c);
    this.receipt = new OriginalReceiptConnector(c);
    this.mandate = new SepaMandateConnector(c);
    this.sepaReport = new SepaReportConnector(c);
    this.trainDelay = new TrainSegmentDelayConnector(c);
    this.routeTemplate = new RouteTemplateConnector(c);
    this.s3blob = opts?.s3 ?? new S3BlobConnector();
  }

  async deleteUser(email: string): Promise<Result<null>> {
    const r = await this.user._query({ KeyConditionExpression: "pk = :v", ExpressionAttributeValues: { ":v": userPk(email) } });
    if (r.isErr()) return r;
    const keys: { pk: string; sk: string }[] = r.value.map((i) => ({ pk: i["pk"] as string, sk: i["sk"] as string }));
    r.value.forEach((i) => {
      const sk = i["sk"] as string;
      if (isPlainTicketSk(sk)) {
        keys.push({ pk: `TICKET#${sk.slice("TICKET#".length)}`, sk: "OWNER" });
      }
    });
    return this.user._batchDelete(keys);
  }

  async deleteTicket(email: string, ticketId: string): Promise<Result<null>> {
    const norm = userPk(email);
    const r = await this.ticket._query({
      KeyConditionExpression: "pk = :pk AND begins_with(sk, :prefix)",
      ExpressionAttributeValues: { ":pk": norm, ":prefix": `TICKET#${ticketId}#` },
    });
    if (r.isErr()) return r;
    const keys: { pk: string; sk: string }[] = r.value.map((i) => ({ pk: i["pk"] as string, sk: i["sk"] as string }));
    keys.push(
      { pk: norm, sk: `TICKET#${ticketId}` },
      { pk: `TICKET#${ticketId}`, sk: "OWNER" },
      { pk: norm, sk: `RAW#${ticketId}` },
      { pk: norm, sk: `RENDERED#${ticketId}` },
    );
    return this.ticket._batchDelete(keys);
  }

  deleteAdmin(email: string) { return this.admin._delete(adminPk(email), "PROFILE"); }
  deleteTicketOwner(ticketId: string) { return this.ticketOwner._delete(`TICKET#${ticketId}`, "OWNER"); }
  deleteRawUpload(email: string, ticketId: string) { return this.rawUpload._delete(userPk(email), `RAW#${ticketId}`); }
  deleteRenderedPdf(email: string, ticketId: string) { return this.renderedPdf._delete(userPk(email), `RENDERED#${ticketId}`); }
  deleteReceipt(email: string, ticketId: string, belegId: string) { return this.receipt._delete(userPk(email), `TICKET#${ticketId}#BELEG#${belegId}`); }
  deleteMandate(email: string, ticketId: string) { return this.mandate._delete(userPk(email), `TICKET#${ticketId}#MANDATE`); }
  deleteSepaReport(date: string, reportId: string) { return this.sepaReport._delete(`SEPA#REPORT#${date}`, `REPORT#${reportId}`); }
  deleteTrainDelay(trainNr: string, date: string, segId: string) { return this.trainDelay._delete(`TRAIN#${trainNr}#${date}`, `SEG#${segId}`); }
  deleteRouteTemplate(email: string, templateId: string) { return this.routeTemplate._delete(userPk(email), `TEMPLATE#${templateId}`); }
}

export { ConflictError };
export { S3BlobConnector };
